{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import os.path as op\n",
    "import argparse\n",
    "import numpy as np\n",
    "import sys\n",
    "import time\n",
    "from copy import copy\n",
    "import itertools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "b'/Users/vitorpavinato/Documents/Faculty_TempleU_2023/Projects/mss_forward_sim'"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwdb()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "def codonInfo():\n",
    "\n",
    "    stopCodons = ['TAG', 'TAA', 'TGA']\n",
    "    codons ={   \"I\":[\"ATT\", \"ATC\", \"ATA\"],\n",
    "                \"L\":[\"CTT\", \"CTC\", \"CTA\", \"CTG\", \"TTA\", \"TTG\"],\n",
    "                \"V\":[\"GTT\", \"GTC\", \"GTA\", \"GTG\"],\n",
    "                \"F\":[\"TTT\", \"TTC\"],\n",
    "                \"M\":[\"ATG\"],\n",
    "                \"C\":[\"TGT\", \"TGC\"],\n",
    "                \"A\":[\"GCT\", \"GCC\", \"GCA\", \"GCG\"],\n",
    "                \"G\":[\"GGT\", \"GGC\", \"GGA\", \"GGG\"],\n",
    "                \"P\":[\"CCT\", \"CCC\", \"CCA\", \"CCG\"],\n",
    "                \"T\":[\"ACT\", \"ACC\", \"ACA\", \"ACG\"],\n",
    "                \"S\":[\"TCT\", \"TCC\", \"TCA\", \"TCG\", \"AGT\", \"AGC\"],\n",
    "                \"Y\":[\"TAT\", \"TAC\"],\n",
    "                \"W\":[\"TGG\"],\n",
    "                \"Q\":[\"CAA\", \"CAG\"],\n",
    "                \"N\":[\"AAT\", \"AAC\"],\n",
    "                \"H\":[\"CAT\", \"CAC\"],\n",
    "                \"E\":[\"GAA\", \"GAG\"],\n",
    "                \"D\":[\"GAT\", \"GAC\"],\n",
    "                \"K\":[\"AAA\", \"AAG\"],\n",
    "                \"R\":[\"CGT\", \"CGC\", \"CGA\", \"CGG\", \"AGA\", \"AGG\"],\n",
    "                \"STOP\":[\"TAA\", \"TAG\", \"TGA\"]}\n",
    "\n",
    "    aalist = list(codons.keys())\n",
    "    codonlist = []\n",
    "\n",
    "    revCodons  = {}\n",
    "    for aa in aalist:\n",
    "        for ci,cd in enumerate(codons[aa]):\n",
    "            codonlist.append(cd)\n",
    "            revCodons[cd] = aa\n",
    "    \n",
    "    return codons, aalist, codonlist, revCodons,stopCodons"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Testing the behaviour of the codonInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "def makeAncestor(args):\n",
    "    \"\"\"\n",
    "        Try to make an ancestral sequence that is at the equilibrium for preferred and unpreferred codons \n",
    "        1.take allDNA and take a random assortment of codons based on appearing in allDNA\n",
    "        2. replace each codon with a random draw from the the most fit neutral set for that same amino acid\n",
    "        3. replace favored codons randomly by making random draws fro mthe less fit neutral set at the approximate equilibrium rate \n",
    "    \"\"\"\n",
    "    \n",
    "    def identifyRandomGene(alignment_location):\n",
    "        \"\"\"\n",
    "        picks random gene from fasta alignment files\n",
    "        \"\"\"\n",
    "        geneFiles = []\n",
    "        for file in os.listdir(alignment_location):\n",
    "            if file.endswith('fasta'):\n",
    "                geneFiles.append(file)\n",
    "        gene = np.random.choice(geneFiles)\n",
    "        print(\"Picked: \" + gene)\n",
    "        return gene\n",
    "\n",
    "    def readInGeneFile(alignment_location,gene = None):\n",
    "        \"\"\"\n",
    "        given random gene selected, returns the alignment data for that gene\n",
    "        \"\"\"\n",
    "        if gene==None:\n",
    "            gene = identifyRandomGene(alignment_location)\n",
    "        with open(op.join(alignment_location,gene)) as f:\n",
    "            species = {}\n",
    "            dna = ''\n",
    "            spec = ''\n",
    "            for line in f:\n",
    "                if line.startswith('>'):\n",
    "                    assert len(dna) % 3 == 0\n",
    "                    species[spec] = dna\n",
    "                    spec = line.lstrip('>').rstrip('\\n')\n",
    "                    dna = ''\n",
    "                else:\n",
    "                    dna += line.strip('\\n')\n",
    "        return species, gene\n",
    "\n",
    "    def createCodonSequence(alignment_location,gene = None):\n",
    "        \"\"\"\n",
    "        given alignments for randomly selected gene, \n",
    "        it turns all alignments into single strand of DNA excluding codons with missing bps or stop codons\n",
    "        \"\"\"\n",
    "        global stopCodons\n",
    "        species, gene = readInGeneFile(alignment_location,gene = gene)\n",
    "        allDNA = ''\n",
    "        for x in species:\n",
    "            dna = species[x]\n",
    "            codon = ''\n",
    "            if len(dna) > 0:\n",
    "                for bp in dna:\n",
    "                    if len(codon) == 3:\n",
    "                        if '-' in codon:\n",
    "                            codon = ''\n",
    "                            continue\n",
    "                        if codon in stopCodons:\n",
    "                            codon = ''\n",
    "                            continue\n",
    "                        allDNA += codon\n",
    "                        codon = ''\n",
    "                    codon += bp\n",
    "        return allDNA, gene\n",
    "\n",
    "    def getCodonProportions(dna):\n",
    "        \"\"\"\n",
    "        calculate proportion of codons for a given strand of DNA, return dictionary\n",
    "        \"\"\"\n",
    "        codons = {}\n",
    "        codon = ''\n",
    "        total = 0\n",
    "\n",
    "        for bp in dna:\n",
    "            if len(codon) == 3:\n",
    "                if codon in codons.keys():\n",
    "                    codons[codon] += 1\n",
    "                else:\n",
    "                    codons[codon] = 1\n",
    "                codon = ''\n",
    "                total += 1\n",
    "            codon += bp\n",
    "\n",
    "        for key in codons.keys():\n",
    "            codons[key] = codons[key] / total\n",
    "\n",
    "        return codons    \n",
    "\n",
    "    global codondic, aalist, codonlist, revcodondic,stopCodons, SynNeuX,SynSelX\n",
    "\n",
    "    dnaStrand,genefilename = createCodonSequence(args.bacaligndir,gene=args.genename)# if args.genename is None,  then a random gene is picked \n",
    "    props = getCodonProportions(dnaStrand)\n",
    "    ancestor = np.random.choice(list(props.keys()), size = args.aalength, p = list(props.values()))\n",
    "    bestcodonancestor = []\n",
    "    fitness = 1.0 \n",
    "    for codon in ancestor:\n",
    "        aa = revcodondic[codon]\n",
    "        if aa in args.neutralsets:\n",
    "            newcodon = np.random.choice(list(args.neutralsets[aa][0]),replace = True) # random codon in first neutral set\n",
    "            bestcodonancestor.append(newcodon)\n",
    "        else:\n",
    "            bestcodonancestor.append(codon)\n",
    "\n",
    "    expterm = np.exp(2*args.SynSel_s)\n",
    "    expprobs2 = 1/(1+expterm)  # equilibrium freq of less fit codon, approx\n",
    "\n",
    "    ancestor = []\n",
    "    nc = 0\n",
    "    for codon in bestcodonancestor:\n",
    "        aa = revcodondic[codon]\n",
    "        if aa in args.neutralsets and len(args.neutralsets[aa]) ==2 and np.random.random() < expprobs2: # change to a less favored codon\n",
    "            newcodon = np.random.choice(list(args.neutralsets[aa][1]))\n",
    "            fitness *= args.fitnessstructure[codon][newcodon] \n",
    "            ancestor.append(newcodon)\n",
    "            nc += 1\n",
    "        else:\n",
    "            ancestor.append(codon)\n",
    "    return ''.join(ancestor),fitness,genefilename"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "def createSelectedDictionary(args):\n",
    "    \"\"\"\n",
    "\n",
    "    \"\"\"\n",
    "    def readModelFile(mssmodelfilename):\n",
    "        \"\"\"\n",
    "        format:  Any line beginning with a '#' is a comment\n",
    "        each line specifies a neutral set of two or more codons for an amino acid \n",
    "        the line begins with either the one letter code for an amino acid\n",
    "        after that comes a series of one or more codons that are respectively neutral to each other (if there is just one, then it is neutral only with respect to itself)\n",
    "        If all codons for an amino acid are neutral,  then that amino acid has only one line. \n",
    "        If an amino acid has multipe sets of neutral codons,  then each set is given on a separate line. \n",
    "        The first line for an amino acid gives the neutral set that has the highest fitness\n",
    "        The second line for an amino acid gives the neutral set that has the next lower fitness\n",
    "        The third line is lower still.   etc \n",
    "        Example:\n",
    "            # A,E,F,P, S, Y all neutral\n",
    "            # G   GGG selected with respect to others\n",
    "            # I ATT selected with respect to others\n",
    "            # L  4-fold set neutral, 2 fold set neutral,  between them is selected \n",
    "            # R  CGT, CGC :neutral  AND  \"CGA\", \"CGG\", \"AGA\", \"AGG\": neutral,  between them selected\n",
    "            # T  ACT, ACC :neutral    ACA, ACG : neutral     between them selected \n",
    "            # V  GTT, GTC: neutral    GTA, GTG: neutral  between them selected \n",
    "            I ATT\n",
    "            I ATC ATA\n",
    "            L CTT CTA CTG CTC \n",
    "            L TTG TTA\n",
    "            V GTT GTC \n",
    "            V GTA GTG\n",
    "            F TTT TTC\n",
    "            C TGT \n",
    "            C TGC\n",
    "            A GCG GCT GCA GCC\n",
    "            G GGA GGC GGT \n",
    "            G GGG\n",
    "            P CCA CCC CCT CCG\n",
    "            T ACC ACT \n",
    "            T ACG ACA\n",
    "            S AGT TCG TCT AGC TCA TCC\n",
    "            Y TAC TAT\n",
    "            Q CAA \n",
    "            Q CAG\n",
    "            N AAT \n",
    "            N AAC\n",
    "            H CAT \n",
    "            H CAC\n",
    "            E GAA GAG\n",
    "            D GAT \n",
    "            D GAC\n",
    "            K AAA \n",
    "            K AAG\n",
    "            R CGC CGT \n",
    "            R CGA AGA AGG CGG        \n",
    "        \"\"\"\n",
    "        global codonlist\n",
    "        checkcodons = {codon:False for codon in codonlist if codon not in [\"TAA\", \"TAG\", \"TGA\",\"ATG\",\"TGG\"]} #avoid STOP, Met, Trp\n",
    "        temp = open(mssmodelfilename,'r').readlines()\n",
    "        lines = [line for line in temp if line[0] != '#' and len(line) > 1]\n",
    "        d = {}\n",
    "        for line in lines:\n",
    "            ls = line.strip().split()\n",
    "            aa = ls[0]\n",
    "            neutralcodonlist = ls[1:]\n",
    "            for codon in neutralcodonlist:\n",
    "                assert checkcodons[codon] == False,\"codon '{}' specified more than once in MSS model file\".format(codon)\n",
    "                checkcodons[codon] = True\n",
    "            if aa in d:\n",
    "                d[aa].append(ls[1:])\n",
    "            else:\n",
    "                d[aa] = [ls[1:]]\n",
    "        assert len(d)==18,\"not all degenerate amino acids specified in MSS model file\"\n",
    "        assert all(codonpresent for codonpresent in checkcodons.values()), \"missing codons {} in MSS model file\".format([key for key, value in checkcodons.items() if value ==False])\n",
    "        return d\n",
    "\n",
    "\n",
    "    global codondic,codonlist,revcodondic,stopCodons\n",
    "    selectedDict = {}\n",
    "    mutDict = {}\n",
    "    neutralsetDict = readModelFile(args.mssmodelfilename)\n",
    "    \n",
    "    #initialize all with nonsense and missense \n",
    "    selectedDict = {}\n",
    "    mutDict = {}\n",
    "    for c1,codon1 in enumerate(codonlist):\n",
    "        tempaaDict = {}\n",
    "        tempaaMuts = {}\n",
    "        aa1 = revcodondic[codon1]\n",
    "        synCodons = codondic[aa1]            \n",
    "        for c2,codon2 in enumerate(codonlist):\n",
    "            aa2 = revcodondic[codon2]\n",
    "            if codon1 in stopCodons or codon2 in stopCodons:\n",
    "                tempaaDict[codon2] = 0.0  ## stop codon\n",
    "                tempaaMuts[codon2] = STOPX ## stop codon \n",
    "            elif codon2 == codon1:\n",
    "                tempaaDict[codon2] = 1.0  ## same codon exactly\n",
    "                tempaaMuts[codon2] = -10 ## same codon  \n",
    "            elif aa1 != aa2:\n",
    "                tempaaDict[codon2] = args.NonSyn_s_rescaled  # nonsynonmous\n",
    "                tempaaMuts[codon2] = NonSynDelX  # [X, -, - ]   \n",
    "        selectedDict[codon1] = tempaaDict\n",
    "        mutDict[codon1] = tempaaMuts           \n",
    "    #now go through and set all synonymous changes \n",
    "    for aa in neutralsetDict:\n",
    "        for ni in range(len(neutralsetDict[aa])):\n",
    "            codonpairs = list(itertools.combinations(neutralsetDict[aa][ni], 2))\n",
    "            for (c1,c2) in codonpairs:\n",
    "                selectedDict[c1][c2] = selectedDict[c2][c1]  = 1.0  # synonymous but neutral\n",
    "                mutDict[c1][c2] = mutDict[c2][c1] = SynNeuX \n",
    "            if len(neutralsetDict) > 1:\n",
    "                for nj in range(ni+1,len(neutralsetDict[aa])):\n",
    "                    codonpairs = list(itertools.product(neutralsetDict[aa][ni],neutralsetDict[aa][nj]))\n",
    "                    for (c1,c2) in codonpairs:\n",
    "                        selectedDict[c1][c2] = args.SynSelDel_s_rescaled  ## both codons found as nonneutral synonymous pairs\n",
    "                        selectedDict[c2][c1] = args.SynSelFav_s_rescaled\n",
    "                        mutDict[c1][c2] = mutDict[c2][c1] = SynSelX \n",
    "\n",
    "    if args.debug and args.debug.usesubtimeinfo:\n",
    "        codon_substitution_time_info_dict = create_codon_pair_substitution_analysis_structure(selectedDict,mutDict)\n",
    "    else:\n",
    "        codon_substitution_time_info_dict = None\n",
    "    return selectedDict, mutDict, neutralsetDict, codon_substitution_time_info_dict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [],
   "source": [
    "def maketreeshape(numSpecies):\n",
    "    \"\"\"\n",
    "    use fixed phylogeny, given number of species \n",
    "    \"\"\"\n",
    "    if numSpecies == 11:\n",
    "        tree = '((((p1,(p5,p6)),(p4,((p8,p9),p7))),(p3,p10)),(p2,p11));'\n",
    "        split_generations = {1: ['p1', 1, None],\n",
    "                            2: ['p2', 0.0, 'p1'],\n",
    "                            3: ['p3', 0.1475, 'p1'],\n",
    "                            4: ['p4', 0.3268, 'p1'],\n",
    "                            5: ['p11', 0.40, 'p2'],\n",
    "                            6: ['p7', 0.4062, 'p4'],\n",
    "                            7: ['p5', 0.4326, 'p1'],\n",
    "                            8: ['p10', 0.5458, 'p3'],\n",
    "                            9: ['p8', 0.5855, 'p7'],\n",
    "                            10: ['p6', 0.8383, 'p5'],\n",
    "                            11: ['p9', 0.9148, 'p8']}\n",
    "        mean_branches_root_to_tip = 4.091\n",
    "    elif  numSpecies==5:\n",
    "        tree = '((((p1,p5),p4),p3),p2);'\n",
    "        split_generations = {1: ['p1', 1, None],\n",
    "                        2: ['p2', 0.0, 'p1'],\n",
    "                        3: ['p3', 0.25, 'p1'],\n",
    "                        4: ['p4', 0.5, 'p1'],\n",
    "                        5: ['p5', 0.75, 'p1']}\n",
    "        mean_branches_root_to_tip = 2.8\n",
    "    elif numSpecies == 4:\n",
    "        tree = '(((p1,p4),p3),p2);'\n",
    "        split_generations = {1: ['p1', 1, None],\n",
    "                        2: ['p2', 0.0, 'p1'],  #0.3\n",
    "                        3: ['p3', 0.4, 'p1'],\n",
    "                        4: ['p4', 0.8, 'p1']}\n",
    "        mean_branches_root_to_tip = 2.25\n",
    "    else: #numSpecies == 1\n",
    "        tree = '(p1);'\n",
    "        split_generations = {1: ['p1', 1, None]}\n",
    "        mean_branches_root_to_tip = 1\n",
    "\n",
    "    return tree,split_generations,mean_branches_root_to_tip \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_codon_pair_substitution_analysis_structure(selectiondic,mtypedic):\n",
    "    \"\"\"\n",
    "        only used when debugging, requires debug.usesubtimeinfo\n",
    "        make a dict of a dict that will contain generation times for different synonymous substitutions \n",
    "    \"\"\"\n",
    "    global codondic,revcodondic,SynSelX,  SynNeuX\n",
    "    cccs = {}\n",
    "\n",
    "    for codon1 in selectiondic:\n",
    "        for codon2 in selectiondic[codon1]:\n",
    "            if codon1 != codon2  and revcodondic[codon1] == revcodondic[codon2] and mtypedic[codon1][codon2] != STOPX:\n",
    "                aa = revcodondic[codon1]\n",
    "                if mtypedic[codon1][codon2] ==SynNeuX: # NEUTRAL\n",
    "                    status = \"NEUTRAL\"\n",
    "                    holdcodon2 = codon2\n",
    "                    holdcodon1 = codon1 \n",
    "                    if holdcodon1 > holdcodon2:# use alphabetic order, if necessary swap them so codon1 is the alphabetically first \n",
    "                        temp = holdcodon2\n",
    "                        holdcodon2 = holdcodon1\n",
    "                        holdcodon1 = temp \n",
    "                    if holdcodon1 in cccs:\n",
    "                        cccs[holdcodon1][holdcodon2] = [aa,status,0,0,0,0] #status, number of forward subs,  total time of forward subs,  number of backward subs, total time of backward subs\n",
    "                    else:\n",
    "                        cccs[holdcodon1] = {holdcodon2:[aa,status,0,0,0,0]}\n",
    "                else: #SELECTED,  get index positions of both codons,  the one to the left of the other in codondic is the favored one,  put that on the left in this dictionary\n",
    "                    assert  mtypedic[codon1][codon2] == SynSelX\n",
    "                    status = \"SELECTED\"\n",
    "                    index1 = codondic[aa].index(codon1)\n",
    "                    index2 = codondic[aa].index(codon2)\n",
    "                    holdcodon2 = codon2\n",
    "                    holdcodon1 = codon1 \n",
    "                    if index1 > index2: # use index order,  if necessary swap so codon1 is to address cccs directly \n",
    "                        temp = holdcodon2\n",
    "                        holdcodon2 = holdcodon1\n",
    "                        holdcodon1 = temp \n",
    "                    if holdcodon1 in cccs:\n",
    "                        cccs[holdcodon1][holdcodon2] = [aa,status,0,0,0,0] #status, number of forward (unfavored) subs,  total time of forward subs,  number of backward (disfavored) subs, total time of backward subs\n",
    "                    else:\n",
    "                        cccs[holdcodon1] = {holdcodon2:[aa,status,0,0,0,0]}\n",
    "    return cccs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [],
   "source": [
    "def makefastafile(samples, fn):\n",
    "    f = open(fn,'w')\n",
    "    for pop in samples.keys():\n",
    "        f.write('>{}\\n'.format(pop))\n",
    "        f.write(str(samples[pop]) + \"\\n\")\n",
    "    f.close()\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "class debuginfo():\n",
    "    \"\"\"\n",
    "        holds some booleans for switching on/off different types of debugging\n",
    "    \"\"\"\n",
    "    def __init__(self,useturnoff,usesubtimeinfo,userescaler):\n",
    "        self.useturnoff = useturnoff\n",
    "        self.usesubtimeinfo = usesubtimeinfo\n",
    "        self.userescaler = userescaler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "class chromosome():\n",
    "    \"\"\"\n",
    "    a chromosome is an 'individual' in the population\n",
    "    it has a DNA sequence from which its fitness is determined \n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self,sequence,fitness,args,mcounts,ancestornumber,fits = None,subtimeinfo=None):\n",
    "        \"\"\"\n",
    "        mcounts :  nummutationtypes positions \n",
    "            0,1,2,3 or 4  for nonsynonymous deleterious, nonsynonymous favored,  synonymous-selected, synonymous-neutral, and STOP \n",
    "        re ancestornumber:\n",
    "          is just the index of the chromosome in the list\n",
    "          it is reset at the beginning of burn2\n",
    "          during burn2 ancestornumber is updated for each chromosome, by copying the value from the ancestor the chromosome was copied from\n",
    "          when all values of ancestornumber are the same,  all the chromosomes are descended from that ancestor in that generation\n",
    "        re subtimeinfo\n",
    "            a list of values about the mutations that accumulate on a chromosome\n",
    "            for each mutation there are 5 values, in order:\n",
    "                codon position\n",
    "                generation number\n",
    "                mutation type\n",
    "                old codon\n",
    "                new codon\n",
    "            these are all appended as they occur,  so the mutations are ordered in time,  and the full list is a length that is a multiple of 5\n",
    "        \"\"\"\n",
    "        global nummutationtypes\n",
    "        self.s = sequence\n",
    "        self.fitstruct = args.fitnessstructure\n",
    "        self.mutstruct = args.mutstructure\n",
    "        self.mrate = args.mutrate\n",
    "        self.mrateinverse = 1.0/self.mrate  # use with exponential random variables for mutation locations because numpy.random.exponential() takes inverse of the rate as the parameter for some reason\n",
    "        self.fitness = fitness\n",
    "        self.mcounts = copy(mcounts)\n",
    "        self.ancestornumber = ancestornumber\n",
    "        self.SynSelFav_s_rescaled = args.SynSelFav_s_rescaled\n",
    "        if args.debug and args.debug.usesubtimeinfo:\n",
    "            if subtimeinfo == None:\n",
    "                self.subtimeinfo = []\n",
    "            else:\n",
    "                self.subtimeinfo = copy(subtimeinfo)\n",
    "            if fits is None:\n",
    "                self.fits = np.empty(shape=0,dtype=np.float128)\n",
    "            else:\n",
    "                self.fits = np.copy(fits)                \n",
    "        else:\n",
    "            self.fits = None\n",
    "    \n",
    "    def chromosomecopy(self,args):\n",
    "        if args.debug and args.debug.usesubtimeinfo:\n",
    "            newchromosome = chromosome(self.s,self.fitness,args,self.mcounts,self.ancestornumber,fits=self.fits, subtimeinfo = self.subtimeinfo)\n",
    "        else:\n",
    "            newchromosome = chromosome(self.s,self.fitness,args,self.mcounts,self.ancestornumber)\n",
    "        return newchromosome\n",
    "\n",
    "    def mutate(self,popancestor,gennum,adaptivechanges,turnoffmut,debug):\n",
    "        \"\"\"\n",
    "            a function that changes the sequence s and recalculates fitness\n",
    "            it uses exponential to get the distance to the next mutation as an approximation for geometric\n",
    "            sites are selected by jumping along the chromosome (distance sampled from exponential)\n",
    "            fitness is not updated until all of the sites in a codon have been changed\n",
    "                usually each mutation is in its own codon, but sometimes 2 or 3 changes occur in the same codon\n",
    "                code is a kind of kludgy, in order to keep a list of all changes in the current codon\n",
    "            \n",
    "        \"\"\"\n",
    "        global mutationlocations # use in debug mode \n",
    "        global  NonSynDelX,  NonSynFavX, SynSelX,  SynNeuX, STOPX \n",
    "        global revcodondic\n",
    "        global args\n",
    "        pos = 0 # a position that mutates  (if not past the end of the sequence)\t\n",
    "        lastcodonpos = -1\n",
    "        muttype = -1\n",
    "        oldCodon = ''\n",
    "        anc = ''\n",
    "        newCodon = ''\n",
    "        fitratio = 0.0\n",
    "        while True:\n",
    "            # distance_to_mut = np.random.geometric(self.mrate) # geometric a bit slower than exponential\n",
    "            while True: # gets one or more changes in a codon\n",
    "                # exponential is faster than geometric,  but can return 0 \n",
    "                distance_to_mut = int(np.random.exponential(self.mrateinverse)) \n",
    "            ## set position that mutates\n",
    "                pos += distance_to_mut\n",
    "                codonpos = pos //3\n",
    "                if lastcodonpos == -1:\n",
    "                    mutlocs = [pos]\n",
    "                    lastcodonpos = codonpos\n",
    "                    pos += 1 # increment to value that is next possible position to mutate \n",
    "                elif codonpos != lastcodonpos:\n",
    "                    lastcodonpos = codonpos\n",
    "                    pos += 1 # increment to value that is next possible position to mutate \n",
    "                    break\n",
    "                else:\n",
    "                    mutlocs.append(pos)\n",
    "                    pos += 1 # increment to value that is next possible position to mutate \n",
    "            if mutlocs[-1] < len(self.s):\n",
    "                ## identify old codon\n",
    "                holds  = self.s\n",
    "                codonpos = mutlocs[0] //3\n",
    "                oldCodon = self.getOldCodon(mutlocs[0]) \n",
    "                # assert oldCodon == self.s[3*codonpos:3*codonpos+3]\n",
    "                # assert (len(mutlocs)==1 or (len(mutlocs)==2 and (mutlocs[1] -mutlocs[0] <= 2)  )  or (len(mutlocs)==3 and (mutlocs[2] -mutlocs[0] <= 2) and (mutlocs[0] < mutlocs[1] < mutlocs[2]) ))\n",
    "                for ml in mutlocs:\n",
    "                    bps =['A', 'G', 'C', 'T']\n",
    "                    bps.remove(self.s[ml:ml+1])\n",
    "                    self.s = self.s[:ml] + np.random.choice(bps) + self.s[ml+1:]\n",
    "                ## update fitness\n",
    "                oldfitness = self.fitness\n",
    "                anc,newCodon,muttype = self.fitnessfunction(mutlocs[0], oldCodon,popancestor)\n",
    "                if muttype == NonSynDelX:\n",
    "                    if codonpos in adaptivechanges and revcodondic[newCodon] == adaptivechanges[codonpos]:\n",
    "                        muttype = NonSynFavX\n",
    "                self.mcounts[muttype] += 1\n",
    "                if debug:\n",
    "                    if (turnoffmut == 1 and muttype in (NonSynDelX,NonSynFavX)) or (turnoffmut==2 and muttype in (SynNeuX,SynSelX)):\n",
    "                        self.s = holds\n",
    "                    else:\n",
    "                        for ml in mutlocs:\n",
    "                            mutationlocations[ml] += 1\n",
    "                        if gennum >= 0 and debug and debug.usesubtimeinfo:\n",
    "                            if len(self.subtimeinfo) >= 7:\n",
    "                                assert gennum >= self.subtimeinfo[-6], \"subtimeinfo genum error {}\".format(self.subtimeinfo)\n",
    "                            fitratio = 0.0 if oldfitness <= 0 else self.fitness/oldfitness\n",
    "                            self.subtimeinfo.append(codonpos)\n",
    "                            self.subtimeinfo.append(gennum)\n",
    "                            self.subtimeinfo.append(muttype)\n",
    "                            self.subtimeinfo.append(fitratio)\n",
    "                            self.subtimeinfo.append(anc)\n",
    "                            self.subtimeinfo.append(oldCodon)\n",
    "                            self.subtimeinfo.append(newCodon)\n",
    "                            if oldfitness > 0.0: # if it is zero it wont end up getting sampled \n",
    "                                self.fits = np.append(self.fits,fitratio)\n",
    "                mainmutationcounter[muttype] += 1\n",
    "                if pos > len(self.s):\n",
    "                    break\n",
    "                else: # reset mutlocs to contain only the last exponential jump position (before it was incremented)\n",
    "                    mutlocs = [pos-1] # the next mutation location (pos was previously incremented to the start of the next interval)\n",
    "            else:\n",
    "                break\n",
    "        return\n",
    "        \n",
    "    def resetmcounts(self):\n",
    "        global nummutationtypes\n",
    "        self.mcounts = [0 for i in range(nummutationtypes)]\n",
    "        self.subtimeinfo = []\n",
    "\n",
    "    def fitnessfunction(self, mut,oldcodon,popancestor):\n",
    "        \"\"\"\n",
    "        recalculates fitness based on newCodon and ancestral codon just for mutations\n",
    "        the ancestor was defined as having a fitness of 1\n",
    "        for x  codons  fitness is a product of x values \n",
    "        the ancestor was defined as having a fitness of 1\n",
    "        for a change from oldcodon to newcodon the fitness is updated by multiplying \n",
    "        fitness times the fitness associated with a change from ancestor to the new codon (i.e. the absolute fitness of the new codon at that position)\n",
    "        and by dividing by the fitness associate with a change from the ancestor to the old codon \n",
    "        \"\"\"\n",
    "        global SynSelX, NonSynDelX,NonSynFavX,STOPX \n",
    "        global stopCodons,revcodondic\n",
    "        anc, newSelf = self.findCodon(mut,popancestor)\n",
    "        # assert newSelf != oldcodon\n",
    "        \n",
    "        muttype = self.mutstruct[oldcodon][newSelf]\n",
    "        if muttype == SynSelX:\n",
    "            self.fitness *=  self.fitstruct[oldcodon][newSelf]\n",
    "        elif muttype == NonSynDelX:\n",
    "            if revcodondic[anc] == revcodondic[newSelf]:\n",
    "                # assert revcodondic[anc] != revcodondic[oldcodon]\n",
    "                self.fitness *= 1.0 / self.fitstruct[anc][oldcodon]\n",
    "            elif revcodondic[anc] == revcodondic[oldcodon]:\n",
    "                # assert revcodondic[anc] != revcodondic[newSelf]\n",
    "                self.fitness *= self.fitstruct[anc][newSelf]\n",
    "            else:\n",
    "                revcodondic[anc] not in (revcodondic[oldcodon],revcodondic[newSelf])\n",
    "                self.fitness *= self.fitstruct[anc][newSelf] / self.fitstruct[anc][oldcodon]\n",
    "        elif muttype == STOPX:\n",
    "            self.fitness  = 0.0\n",
    "        return anc,newSelf, muttype\n",
    "\n",
    "    def getOldCodon(self, i):\n",
    "        \"\"\"\n",
    "        identify codon that mutation is in for ancestral and sequence\n",
    "        \"\"\"\n",
    "        position = i % 3\n",
    "\n",
    "        if position == 0:\n",
    "            return self.s[i:i+3]\n",
    "        elif position == 1:\n",
    "            return self.s[i-1:i+2]\n",
    "        elif position == 2:\n",
    "            return self.s[i-2:i+1]\n",
    "        else:\n",
    "            print('getOldCodon error position {} in sequence {}'.format(i,self.s))\n",
    "            exit()\n",
    "\n",
    "    def findCodon(self, i,popancestor):\n",
    "        \"\"\"\n",
    "        identify codon that mutation is in for ancestral and sequence\n",
    "        \"\"\"\n",
    "        position = i % 3\n",
    "\n",
    "        if position == 0:\n",
    "            return popancestor[i:i+3], self.s[i:i+3]\n",
    "        elif position == 1:\n",
    "            return popancestor[i-1:i+2], self.s[i-1:i+2]\n",
    "        elif position == 2:\n",
    "            return popancestor[i-2:i+1], self.s[i-2:i+1]\n",
    "        else:\n",
    "            print('findcodon() error position {} in ancestor {}'.format(i,popancestor))\n",
    "            exit()\n",
    "\n",
    "    def __str__(self):\n",
    "        return self.s\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": [
    "class population(list):\n",
    "    \"\"\"\n",
    "    basically a list of chromosomes with some added functionality\n",
    "\n",
    "    \"\"\"\n",
    "    def __init__(self, label, source, args):\n",
    "        \"\"\"\n",
    "        if source is a sequence, then it is the ancestral sequence and all chromosomes are made as copies of it\n",
    "        if source is a population then the new population is made by sampling from it at random\n",
    "        \"\"\"\n",
    "        global nummutationtypes\n",
    "        self.label = label\n",
    "        self.mrate = args.mutrate\n",
    "        self.popsize2 = args.popsize2\n",
    "        self.args = args\n",
    "        self.popancestor = self.args.ancestor\n",
    "        if isinstance(source,population):\n",
    "            \"\"\"\n",
    "            copy each chromosome from source and put it in the new population\n",
    "            \"\"\"\n",
    "            for chrom in source:\n",
    "                # self.append(chrom)\n",
    "                self.append(chrom.chromosomecopy(args))\n",
    "        else:# at the beginning, fill up the pouplation with chromosomes made from the ancestor \n",
    "            for i in range(self.popsize2):\n",
    "                self.append(chromosome(source,args.ancestorfitness,args, [0 for j in range(nummutationtypes)],i))\n",
    "        self.adaptivechanges = {}\n",
    "\n",
    "    def generation(self,gennum):\n",
    "        \"\"\"\n",
    "        random sampling of the next generation based on the fitnesses of the chromosomes\n",
    "            make array of fitnesses\n",
    "            get list of unique values and indices for these values\n",
    "            get expected frequencies\n",
    "            sample randomparents using multinomial \n",
    "        after each chromosome is sampled,  mutations are added and fitness is recalculated\n",
    "        replace the old population with the new sampled chromosomes \n",
    "        \"\"\"\n",
    "        fits = np.array([c.fitness for c in self],dtype=float)\n",
    "        unique_vals,indices,counts = np.unique(fits, return_counts=True,return_inverse=True)\n",
    "        unique_indices = []\n",
    "        for i in range(len(unique_vals)):\n",
    "            unique_indices.append(np.where(indices == i)[0])\n",
    "\n",
    "        # this is not faster than above 3 lines of code \n",
    "        # unique_vals, inverse_indices,counts = np.unique(fits,return_counts=True, return_inverse=True)\n",
    "        # unique_indices = np.split(np.argsort(inverse_indices), np.cumsum(np.unique(inverse_indices, return_counts=True)[1])[:-1])\n",
    "        numfits = unique_vals.shape[0]\n",
    "        if numfits == 1:\n",
    "            randomparentids = [np.random.choice(unique_indices[0],size = self.args.popsize2,replace = True)]        \n",
    "        else:\n",
    "            expfreqs = (unique_vals*counts)/len(fits)/fits.mean()\n",
    "            expfreqs = [v if v <= 1.0 else 1.0 for v in expfreqs] # can get a value slightly greater than 1  when a fitness is 0 \n",
    "            samplecounts = np.random.multinomial(self.args.popsize2,expfreqs,1)\n",
    "            randomparentids = [np.random.choice(unique_indices[i],size=samplecounts[0,i],replace=True) for i in range(numfits)]\n",
    "            if self.args.debug:\n",
    "                for parentgroup in randomparentids:\n",
    "                    for i in parentgroup:\n",
    "                        assert self[i].fitness > 0 \n",
    "        newpop = []       \n",
    "        for parentgroup in randomparentids:\n",
    "            for i in parentgroup:\n",
    "                #copy the chromosome.  this is much, much faster than deepcopy\n",
    "                child = self[i].chromosomecopy(self.args)\n",
    "                child.mutate(self.popancestor,gennum,self.adaptivechanges,self.args.turnoffmut,self.args.debug)\n",
    "                newpop.append(child)\n",
    "                \n",
    "        self.clear()\n",
    "        for child in newpop:\n",
    "            self.append(child)\n",
    "        return numfits\n",
    "    \n",
    "    def changeancestor(self):\n",
    "        global codondic,revcodondic,aalist\n",
    "        pos = np.random.randint(self.args.aalength)\n",
    "        codon = self.popancestor[3*pos:3*pos+3]\n",
    "        aa = revcodondic[codon]\n",
    "        while True:\n",
    "            newaa = np.random.choice(aalist)\n",
    "            if newaa not in (aa,'STOP'):\n",
    "                break\n",
    "        newcodon = np.random.choice(codondic[newaa])\n",
    "        self.popancestor = self.popancestor[:3*pos] + newcodon + self.popancestor[3*(pos+1):]\n",
    "        assert len(self.popancestor) == 3*self.args.aalength\n",
    "        for c in self:\n",
    "            if revcodondic[c.s[3*pos:3*pos+3]] != newaa:\n",
    "                c.fitness *= self.args.NonSyn_s_rescaled\n",
    "        self.adaptivechanges[pos] = newaa\n",
    "        return pos,newaa\n",
    "        \n",
    "    def checkancestors(self):\n",
    "        anc0 = self[0].ancestornumber\n",
    "        allthesame = all(anc0 == c.ancestornumber for c in self)\n",
    "        return allthesame\n",
    "    \n",
    "    def sampleindividual(self, num):\n",
    "        \"\"\"\n",
    "        return random chromosomes of number num from population\n",
    "        \"\"\"\n",
    "        return np.random.choice(self, num)\n",
    "\n",
    "    def checkpop(self, seqLen, gen):\n",
    "        \"\"\"\n",
    "            check whatever should be true\n",
    "            e.g. immediately after call to generation()  none should have fitness of 0\n",
    "        \"\"\"\n",
    "          ## make sure the popsize is constant\n",
    "        assert len(self) == self.popsize2\n",
    "\n",
    "        ## make sure the length of chromosome is right - check random chromsosme\n",
    "        assert len(self[np.random.randint(self.popsize2)].s) == seqLen*3\n",
    "\n",
    "    def reset_mutation_counts(self):\n",
    "        for c in self:\n",
    "            c.resetmcounts()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [],
   "source": [
    "class tree():\n",
    "    \"\"\"\n",
    "        Represents a tree of populations\n",
    "        makes initial population\n",
    "        when tree.run() is called it runs the simulation and at the end returns the sample\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self,args):\n",
    "        self.treestr = args.tree\n",
    "        self.args = args\n",
    "        self.split_generations, self.times = self.translateGens(args.split_generations)\n",
    "        self.pop0 = population('p1', args.ancestor,args)\n",
    "        self.pops = {}\n",
    "\n",
    "    def translateGens(self, sg):\n",
    "        newSG = {}\n",
    "        times = []\n",
    "        for key in sorted(sg.keys()):\n",
    "            newTime = round(sg[key][1] * self.args.treeDepth )\n",
    "            newSG[newTime] = [sg[key][0], sg[key][2]]\n",
    "            times.append(newTime)\n",
    "        return newSG, times\n",
    "\n",
    "\n",
    "    def samplefrompops(self):\n",
    "        \"\"\"\n",
    "        samples sequences at the end of the run\n",
    "        \"\"\"\n",
    "        global stopCodons\n",
    "        samples = {}\n",
    "        for pop in self.pops.keys():\n",
    "            while True: # avoid sampling an individual with a stop codon, will hang if all individuals have a stop codon\n",
    "                temp =self.pops[pop].sampleindividual(1)[0]\n",
    "                notok = True in [codon in stopCodons for codon in [temp.s[i:i+3] for i in range(0, len(temp.s), 3)]]\n",
    "                if notok is False:\n",
    "                    break\n",
    "            samples[pop] = temp\n",
    "        return samples\n",
    "    \n",
    "    # New function to output population-level alignments\n",
    "    def sampleindsfromonepop(self, nindiv, pop):\n",
    "        \"\"\"\n",
    "        samples n sequences(individuals) from a population at the end of the run (for SFS calculation).\n",
    "        It saves genomes from a randomly picked population to a dictionary of ind:sequences.\n",
    "        \"\"\"\n",
    "        if nindiv > 0:\n",
    "            # Raise errors \n",
    "            if pop > self.args.numSpecies:\n",
    "                raise ArithmeticError(f\"the chosen population p{pop} does not exists!\")\n",
    "\n",
    "            # Question here: all populations have the same self.args.popsize?\n",
    "            if nindiv > self.args.popsize:\n",
    "                raise AttributeError(f\"n = {nindiv} is not valid as n > N!\")\n",
    "        \n",
    "            # Define the number of sampled chromosomes based on n\n",
    "            nchrom = 2*nindiv\n",
    "        \n",
    "            # Define how to select a population to sample from\n",
    "            if pop == 0:\n",
    "                # sample a random population\n",
    "                popkey = np.random.choice(list(sim.pops.keys()))\n",
    "            else:\n",
    "                popkey = \"p\" + str(pop)\n",
    "\n",
    "            # print(f\"available populations to sample from are: {sim.pops.keys()}\")\n",
    "            # print(f\"selected pop is {popkey}\")\n",
    "\n",
    "            # Sample from the population\n",
    "            sampled_seqs = self.pops[popkey].sampleindividual(nchrom)\n",
    "        \n",
    "            # create a list of keys\n",
    "            c = list(range(0, len(sampled_seqs)))\n",
    "            seqnames = [s + str(i) + '_' + popkey for i, s in zip(c, itertools.cycle(\"c\"))]\n",
    "        \n",
    "            # create a dictionary of indiv Ids as keys and sequences as values    \n",
    "            sampled_chroms = {}\n",
    "            for i, key in enumerate(seqnames):\n",
    "                sampled_chroms[key] = sampled_seqs[i]\n",
    "\n",
    "            # This will save a different fasta file with population-level alignments\n",
    "            indvfastafilename = op.join(self.args.fdir, self.args.basename +  \"_\" + self.args.genename +\"_nindivfrompop.fa\")\n",
    "            if os.path.exists(indvfastafilename):\n",
    "                indvfastafilename = '{}(1).fa'.format(indvfastafilename[:-3])\n",
    "            \n",
    "            # This does the magic\n",
    "            makefastafile(sampled_chroms, indvfastafilename)\n",
    "            \n",
    "\n",
    "    def fitCheck(self):\n",
    "        \"\"\"\n",
    "        picka  random chromosome from the population \n",
    "        #write fitnesses to log file\n",
    "        return mean fitness\n",
    "        \"\"\"\n",
    "        meanfits = []\n",
    "        popkeys = self.pops.keys()\n",
    "        for pop in popkeys:\n",
    "            meanfits.append(sum([c.fitness for c in self.pops[pop]])/len(self.pops[pop]))\n",
    "        return meanfits,sum(meanfits)/len(meanfits)\n",
    "\n",
    "    def fitmutsummary(self):\n",
    "        \"\"\"\n",
    "        pick a  random chromosome from each population\n",
    "        return a list  [meanfitness, list of fitness,  list of mcount lists]\n",
    "        \"\"\"\n",
    "        meanfit = 0\n",
    "        popkeys = self.pops.keys()\n",
    "        fitlist = []\n",
    "        mcountlist = []\n",
    "        assert len(popkeys)== self.args.numSpecies \n",
    "        for pop in popkeys:\n",
    "            num = np.random.randint(self.args.popsize2)\n",
    "            temp = self.pops[pop][num].fitness\n",
    "            meanfit += temp\n",
    "            fitlist.append(temp)\n",
    "            mcountlist.append(self.pops[pop][num].mcounts)\n",
    "        return meanfit/len(popkeys),fitlist,mcountlist\n",
    "    \n",
    "    def checkcodonsubstitutions(self,sampledsequences,cccs):\n",
    "            global nummutationtypes\n",
    "            for p in sampledsequences:\n",
    "                c = sampledsequences[p]\n",
    "                d = {}\n",
    "                prevgens = 0\n",
    "                for i in range(0, len(c.subtimeinfo), 7):\n",
    "                    pos = c.subtimeinfo[i]\n",
    "                    sub = [c.subtimeinfo[i], c.subtimeinfo[i + 5], c.subtimeinfo[i + 6]]  \n",
    "                    gens = c.subtimeinfo[i+1]\n",
    "                    assert gens >= 0, \"{} {}\".format(gens, c.subtimeinfo[i:i+7])\n",
    "                    muttype = c.subtimeinfo[i+2]\n",
    "                    codon1 = sub[1]\n",
    "                    codon2 = sub[2]         \n",
    "                    k = i\n",
    "                    while True:\n",
    "                        k -= 5\n",
    "                        if k >= 0:\n",
    "                            if c.subtimeinfo[k] == pos:\n",
    "                                prevcodon1 = c.subtimeinfo[k+5]\n",
    "                                prevcodon2 = c.subtimeinfo[k+6]\n",
    "                                prevgens = c.subtimeinfo[k+1]\n",
    "                                useit = True\n",
    "                                break\n",
    "                        else:\n",
    "                            useit = False\n",
    "                            break\n",
    "                    if useit and  codon1 == prevcodon2 and codon2 == prevcodon1:\n",
    "                        if codon1 in cccs and codon2 in cccs[codon1]:\n",
    "                            cccs[codon1][codon2][2] += 1\n",
    "                            cccs[codon1][codon2][3] += gens - prevgens\n",
    "                        elif codon2 in cccs and codon1 in cccs[codon2]:\n",
    "                            cccs[codon2][codon1][4] += 1\n",
    "                            cccs[codon2][codon1][5] += gens - prevgens\n",
    "                        if sub[0] in d:\n",
    "                            d[sub[0]].append(codon1)\n",
    "                            d[sub[0]].append(codon2)\n",
    "                        else:\n",
    "                            d[sub[0]] = [codon1,codon2]\n",
    "                for pos in d: #need to debug why this crashes when this is on when debugging \n",
    "                    if len(d[pos])>3:\n",
    "                        for k in range(1, len(d[pos]), 2):\n",
    "                            if k <  len(d[pos])-1:\n",
    "                                assert d[pos][k] == d[pos][k+1],\"{} {} {} {}\".format(i,pos,d[pos][k],d[pos][k+1] )\n",
    "            return cccs\n",
    "    \n",
    "    def makesubtimeinfoliststring(self,sampledsequences):\n",
    "        global revcodondic,codondic\n",
    "        s = [\"\\npop#\\tAApos\\tgennum\\ttype\\tfitratio\\tancAA\\tancCodon\\tfromAA\\tfromCodon\\tfromIndex\\ttoAA\\ttoCodon\\ttoIndex\\n\"] \n",
    "        for pi,p in enumerate(sampledsequences):\n",
    "            c = sampledsequences[p]\n",
    "            i = 0\n",
    "            subtimeinfo = [c.subtimeinfo[i:i + 7] for i in range(0, len(c.subtimeinfo), 7)]\n",
    "            for sti in subtimeinfo:\n",
    "                s.append(\"{}\\t{}\\t{}\\t{}\\t{:.5f}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n\".format(pi,sti[0],sti[1],sti[2],sti[3],revcodondic[sti[4]],sti[4],revcodondic[sti[5]],sti[5],codondic[revcodondic[sti[5]]].index(sti[5]),revcodondic[sti[6]],sti[6],codondic[revcodondic[sti[6]]].index(sti[6])))\n",
    "        return ''.join(s)\n",
    "    \n",
    "    def calc_fitness_from_changes(self,sampledsequences):\n",
    "        fitness_from_changes = []\n",
    "        numchanges = [len(sampledsequences[p].fits)-1 for p in sampledsequences]\n",
    "        for p in sampledsequences:\n",
    "            c = sampledsequences[p]\n",
    "            fitness_from_changes.append(np.prod(c.fits))\n",
    "        return fitness_from_changes,numchanges\n",
    "\n",
    "    def makefitchangeinfostring(self,sampledsequences):\n",
    "        global revcodondic,codondic\n",
    "        s = [\"\\npop#\\tsub#\\tfitchange\\tAApos\\tgen\\ttype\\tfromAA\\tfromCodon\\tfromIndex\\ttoAA\\ttoCodon\\ttoIndex\\n\"] \n",
    "        for pi,p in enumerate(sampledsequences):\n",
    "            c = sampledsequences[p]\n",
    "            i = 0\n",
    "            subtimeinfo = [c.subtimeinfo[i:i + 5] for i in range(0, len(c.subtimeinfo), 5)]\n",
    "            for f,m in zip(c.fits[1:],subtimeinfo):\n",
    "                if True:#f> 1:\n",
    "                    s.append(\"{}\\t{}\\t{:.5f}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\t{}\\n\".format(pi,i,f,m[0],m[1],m[2],revcodondic[m[3]],m[3],codondic[revcodondic[m[3]]].index(m[3]),revcodondic[m[4]],m[4],codondic[revcodondic[m[4]]].index(m[4])))\n",
    "                i += 1\n",
    "        return ''.join(s)\n",
    "    \n",
    "    def makeancestoraminoacidstring(self):\n",
    "        global revcodondic,aalist\n",
    "        sum = 0\n",
    "        aacountdic = {aa:0 for aa in aalist}\n",
    "        codons = [self.args.ancestor[i:i+3] for i in range(0,len(self.args.ancestor),3)]\n",
    "        for codon in codons:\n",
    "            aacountdic[revcodondic[codon]] += 1\n",
    "            sum += 1\n",
    "        ancestoraminoacidstring  = [\"\\nAncestor Amino Acid Counts and Frequencies:\\n\"]\n",
    "        for aa in aacountdic:\n",
    "            ancestoraminoacidstring.append (\"{}\\t{}\\t{:.3f}\\n\".format(aa,aacountdic[aa],aacountdic[aa]/sum))\n",
    "        ancestoraminoacidstring.append(\"\\n\")\n",
    "        return ''.join(ancestoraminoacidstring)\n",
    "\n",
    "        \n",
    "    def write_debug_information(self,sampledsequences,rf):\n",
    "        rf.write(\"\\n\\nDEBUGGING STUFF\\n\\n\")\n",
    "\n",
    "        fitness_from_changes,numchanges = self.calc_fitness_from_changes(sampledsequences)\n",
    "        rf.write(\"\\nSample chromosome fitnesses calculated from changes\\n\")\n",
    "        rf.write(\"Pop\\t#changes\\tfitness\\n\")\n",
    "        for i in range(len(numchanges)):\n",
    "            rf.write(\"{}\\t{}\\t{:.10f}\\n\".format(i,numchanges,fitness_from_changes[i]))\n",
    "        self.args.codon_substitution_time_info_dict = self.checkcodonsubstitutions(sampledsequences,self.args.codon_substitution_time_info_dict)\n",
    "        rf.write(\"\\tMean Synonymous Substitution Times: NEUTRAL Pairs\\n\\nAA\\tCodon1\\tCodon2\\t#Frwd\\tMeanFrwdGens\\t#Bkwd\\tMeanBkwdGens\\n\")\n",
    "        fc=bc=fsum = bsum = 0\n",
    "        for codon1 in self.args.codon_substitution_time_info_dict:\n",
    "            for codon2 in self.args.codon_substitution_time_info_dict[codon1]:\n",
    "                x = self.args.codon_substitution_time_info_dict[codon1][codon2] # fewer characters\n",
    "                if x[1] == \"NEUTRAL\":\n",
    "                    fc += x[2]\n",
    "                    fsum += x[3]\n",
    "                    bc += x[4]\n",
    "                    bsum += x[5]\n",
    "        rf.write(\"Overall\\t#Frwd\\tMeanFrwdGens\\t#Bkwd\\tMeanBkwdGens\\n\")\n",
    "        rf.write(\"mean   \\t{}\\t{:.0f}\\t{}\\t{:.0f}\\n\".format(fc,np.nan if fc==0 else fsum/fc,bc,np.nan if bc == 0 else bsum/bc))\n",
    "        for codon1 in self.args.codon_substitution_time_info_dict:\n",
    "            for codon2 in self.args.codon_substitution_time_info_dict[codon1]:\n",
    "                x = self.args.codon_substitution_time_info_dict[codon1][codon2] # fewer characters\n",
    "                if x[1] == \"NEUTRAL\":\n",
    "                    rf.write(\"{}\\t{}\\t{}\\t{}\\t{:.0f}\\t{}\\t{:.0f}\\n\".format(x[0],codon1,codon2,x[2],np.nan if x[2]==0 else x[3]/x[2],x[4],np.nan if x[4]==0 else x[5]/x[4]))\n",
    "        rf.write(\"\\nMean Synonymous Substitution Times: SELECTED Pairs\\n\\nAA\\tCodon1\\tCodon2\\t#Frwd\\tMeanFrwdGens\\t#Bkwd\\tMeanBkwdGens\\n\")\n",
    "        fc=bc=fsum = bsum = 0\n",
    "        for codon1 in self.args.codon_substitution_time_info_dict:\n",
    "            for codon2 in self.args.codon_substitution_time_info_dict[codon1]:\n",
    "                x = self.args.codon_substitution_time_info_dict[codon1][codon2] # fewer characters\n",
    "                if x[1] == \"SELECTED\":\n",
    "                    fc += x[2]\n",
    "                    fsum += x[3]\n",
    "                    bc += x[4]\n",
    "                    bsum += x[5]\n",
    "        rf.write(\"Overall\\t#Frwd\\tMeanFrwdGens\\t#Bkwd\\tMeanBkwdGens\\n\")\n",
    "        rf.write(\"mean   \\t{}\\t{:.0f}\\t{}\\t{:.0f}\\n\".format(fc,np.nan if fc==0 else fsum/fc,bc,np.nan if bc == 0 else bsum/bc))\n",
    "        for codon1 in self.args.codon_substitution_time_info_dict:\n",
    "            for codon2 in self.args.codon_substitution_time_info_dict[codon1]:\n",
    "                x = self.args.codon_substitution_time_info_dict[codon1][codon2] # fewer characters\n",
    "                if x[1] == \"SELECTED\":\n",
    "                    rf.write(\"{}\\t{}\\t{}\\t{}\\t{:.0f}\\t{}\\t{:.0f}\\n\".format(x[0],codon1,codon2,x[2],np.nan if x[2]==0 else x[3]/x[2],x[4],np.nan if x[4]==0 else x[5]/x[4]))\n",
    "        ancestoraminoacidstring = self.makeancestoraminoacidstring()\n",
    "        rf.write(ancestoraminoacidstring)\n",
    "        # changeinfostring = self.makefitchangeinfostring(sampledsequences)\n",
    "        # rf.write(changeinfostring)\n",
    "        subtimeinfostring = self.makesubtimeinfoliststring(sampledsequences)\n",
    "        rf.write(subtimeinfostring)\n",
    "        return rf\n",
    "\n",
    "    def summarize_results(self,starttime,sampledsequences):\n",
    "\n",
    "        class   substitution_info():\n",
    "            def __init__(self,label, count,subs,totalcount,totalnumgen,args,parent=None):\n",
    "                if parent is not None:\n",
    "                    self.label = parent + \"_\" + label\n",
    "                else:\n",
    "                    self.label = label\n",
    "                while len(self.label) < 18:\n",
    "                    self.label += ' '\n",
    "                self.count = count \n",
    "                self.subs = subs\n",
    "                self.mutproportion = count/totalcount\n",
    "                self.ebp = 3*args.aalength*self.mutproportion\n",
    "                self.mutperebp = np.nan if self.ebp == 0 else self.count/self.ebp\n",
    "                self.subrate = np.nan if self.ebp == 0 else self.subs/self.ebp\n",
    "                self.totalnumgen = totalnumgen\n",
    "                self.args = args\n",
    "\n",
    "            def tablestr(self,tabletype,withheader):\n",
    "                if withheader == False:\n",
    "                    header = \"\"\n",
    "                else:\n",
    "                    if tabletype == \"Mutation\":\n",
    "                        header = \"\\nMutation Total Counts/Rates\\n\\tmutation_type    total_count  effective_#bp  proportions\tmutations_per_effective_bp:\\n\"\n",
    "                    else: # tabletype == \"Substitution\"\n",
    "                        header = \"\\nSubstitution Counts/Rates\\n\\tsubstitution_type\tcount_per_gene\tper_effective_bp\tper_effective_bp_per_branch\tper_effective_bp_per_generation\\n\"\n",
    "                if tabletype == \"Mutation\":\n",
    "                    return \"{}\\t{}\\t{:>12d}\\t{:>3.1f}\\t{:>6.3g}\\t{:.3g}\\n\".format(header,self.label,self.count,self.ebp,self.mutproportion,self.mutperebp)\n",
    "                else:\n",
    "                    assert tabletype == \"Substitution\"\n",
    "                    return \"{}\\t{}\\t{:>4.1f}\\t{:>4.3g}\\t{:>3.3g}\\t{:.3g}\\n\".format(header,self.label,self.subs,self.subrate,self.subrate/self.args.mean_branches_root_to_tip ,self.subrate/self.totalnumgen)\n",
    "\n",
    "            def ratiostr(self,denominator,label,withheader=False):\n",
    "                if withheader == False:\n",
    "                    header = \"\"\n",
    "                else:\n",
    "                    header = \"\\nRate Ratios:\\n\"\n",
    "                while len(label) < 65:\n",
    "                    label += ' '\n",
    "                ratio = np.nan if denominator.subrate == 0 else self.subrate/denominator.subrate \n",
    "                return \"{}\\t{}\\t{:.3g}\\n\".format(header,label,ratio)\n",
    "                        \n",
    "        global mainmutationcounter\n",
    "        global nummutationtypes\n",
    "        global NonSynDelX, NonSynFavX, SynSelX, SynNeuX, STOPX\n",
    "        mnames = [\"NonSyn_Del\",\"NonSyn_Fav\",\"Synon_Sel\",\"Synon_Neu\",\"Stop\"]\n",
    "        meanfit,fitlist,mcountlist = self.fitmutsummary()\n",
    "        rf = open(self.args.resultsfilename,'w')\n",
    "        rf.write(\"mss_sim\\n\\narguments:\\n\")\n",
    "        for arg in vars(self.args):\n",
    "            if arg==\"fitnessstructure\" or  arg==\"mutstructure\" or arg == \"codon_substitution_time_info_dict\":\n",
    "                if self.args.debug:\n",
    "                    rf.write(\"\\t{}: {}\\n\".format(arg, getattr(self.args, arg)))\n",
    "                else:\n",
    "                    rf.write(\"\\t{}: {}\\n\".format(arg, \" - printed only in debug mode\")) # quite large \n",
    "            else:\n",
    "                rf.write(\"\\t{}: {}\\n\".format(arg, getattr(self.args, arg)))\n",
    "\n",
    "        rf.write(\"\\nFinal Mean Fitness: {:.4g}\\n\".format(meanfit))\n",
    "        rf.write(\"\\nMean number of fitness values each generation: {:.1f}\\nMean number of individuals per fitness value: {:.1f}\\n\".format(self.args.meannumfits,self.args.popsize2/self.args.meannumfits))\n",
    "        rf.write(\"\\nSampled Individual Fitnesses: {}\\n\".format(fitlist))\n",
    "        rf.write(\"\\nSampled Individual Mutation Counts ({}): {}\\n\".format(mnames,mcountlist))  \n",
    "\n",
    "        totalnumgen = self.args.treeDepth + self.args.burn2_generation_time\n",
    "        allmuttotsum = sum(mainmutationcounter)\n",
    "        subsum = [0 for i in range(nummutationtypes)]\n",
    "        for mc in mcountlist:\n",
    "            for i in range(nummutationtypes):\n",
    "                subsum[i] += mc[i]\n",
    "        for i in range(nummutationtypes): # take the mean count per sampled chromosome\n",
    "            subsum[i] /= self.args.numSpecies\n",
    "\n",
    "        #populate substitution info list\n",
    "        \n",
    "        # relies on global vars, in order  NonSynDelX, NonSynFavX, SynSelX, SynNeuX, STOPX\n",
    "        subinfolist = []\n",
    "        # all nonsynonymous subinfolist[0]\n",
    "        allmuts = mainmutationcounter[NonSynDelX] + mainmutationcounter[NonSynFavX]\n",
    "        meansubs = subsum[NonSynDelX] + subsum[NonSynFavX]\n",
    "        subinfolist.append(substitution_info(\"NonSyn\",allmuts,meansubs,allmuttotsum,totalnumgen,self.args))\n",
    "        # nonsynonymous deleterious subinfolist[1]\n",
    "        subinfolist.append(substitution_info(\"Deleterious\",mainmutationcounter[NonSynDelX],subsum[NonSynDelX],allmuttotsum,totalnumgen,self.args,parent=\"NonSyn\"))\n",
    "        # nonsynonymous favored subinfolist[2]\n",
    "        subinfolist.append(substitution_info(\"Favored\",mainmutationcounter[NonSynFavX],subsum[NonSynFavX],allmuttotsum,totalnumgen,self.args,parent=\"NonSyn\"))\n",
    "        # all synonymous subinfolist[3]\n",
    "        allmuts = mainmutationcounter[SynSelX] + mainmutationcounter[SynNeuX]\n",
    "        meansubs = subsum[SynSelX] + subsum[SynNeuX]\n",
    "        subinfolist.append(substitution_info(\"Synon\",allmuts,meansubs,allmuttotsum,totalnumgen,self.args))\n",
    "        # synonymous selected  subinfolist[4]\n",
    "        subinfolist.append(substitution_info(\"Selected\",mainmutationcounter[SynSelX],subsum[SynSelX],allmuttotsum,totalnumgen,self.args,parent=\"Synon\"))\n",
    "        # synonymous neutral subinfolist[5]\n",
    "        subinfolist.append(substitution_info(\"Neutral\",mainmutationcounter[SynNeuX],subsum[SynNeuX],allmuttotsum,totalnumgen,self.args,parent=\"Synon\"))\n",
    "        # STOP  subinfolist[6]\n",
    "        subinfolist.append(substitution_info(\"STOP\",mainmutationcounter[STOPX],subsum[STOPX],allmuttotsum,totalnumgen,self.args))\n",
    "\n",
    "        #print tables\n",
    "        tabletype = \"Mutation\"\n",
    "        for i,sb in enumerate(subinfolist):\n",
    "            rf.write(sb.tablestr(tabletype,i==0))\n",
    "        tabletype = \"Substitution\"\n",
    "        for i,sb in enumerate(subinfolist):\n",
    "            rf.write(sb.tablestr(tabletype,i==0))\n",
    "        #rate ratios\n",
    "        rf.write(subinfolist[1].ratiostr(subinfolist[3],\"\\tdN*/dS (Nonsynonymous_deleterious/Synonymous (selected and neutral)\",withheader=True))\n",
    "        rf.write(subinfolist[0].ratiostr(subinfolist[3],\"\\tdN/dS (Nonsynonymous_total/Synonymous (selected and neutral)\"))\n",
    "        rf.write(subinfolist[1].ratiostr(subinfolist[5],\"\\tdN*/dSn (Nonsynonymous_deleterious/Synonymous_Neu)\"))\n",
    "        rf.write(subinfolist[0].ratiostr(subinfolist[5],\"\\tdN/dSn (Nonsynonymous_total/Synonymous_Neu)\"))\n",
    "        rf.write(subinfolist[4].ratiostr(subinfolist[5],\"\\tdSs/dSn (Synonymous_Sel/Synonymous_Neu)\"))\n",
    "        if self.args.debug and self.args.debug.usesubtimeinfo:\n",
    "            rf = self.write_debug_information(sampledsequences,rf)\n",
    "                \n",
    "        totaltime = time.time()-starttime\n",
    "        rf.write(\"\\ntotal time: {}\\n\".format(time.strftime(\"%H:%M:%S\",time.gmtime(totaltime))))            \n",
    "\n",
    "        rf.close()\n",
    "        \n",
    "\n",
    "    def run_burn1(self):\n",
    "        \"\"\"\n",
    "        run for a multiplier,  burn1X, x popsize2 generations\n",
    "        200x  seems good enough,  with the 5/5/ version of makeAncestor()\n",
    "        \"\"\"\n",
    "        defaultgens = 200\n",
    "        if self.args.debug and self.args.debug.userescaler:\n",
    "                burn1X = int(defaultgens * self.args.treerescaler) \n",
    "        else:\n",
    "               burn1X = defaultgens\n",
    "        for i in range(burn1X*self.args.popsize2):\n",
    "            self.pop0.generation(-1)\n",
    "            if self.args.debug and  i % 1000 == 0:\n",
    "                    meanfit = sum([c.fitness for c in self.pop0])/len(self.pop0)\n",
    "                    print(i, meanfit)                \n",
    "        meanfit = sum([c.fitness for c in self.pop0])/len(self.pop0)\n",
    "        self.pop0.reset_mutation_counts()  \n",
    "        if self.args.debug and self.args.debug.usesubtimeinfo:\n",
    "            for c in self.pop0:\n",
    "                c.fits = [c.fitness]\n",
    "        \n",
    "        return i+1,meanfit\n",
    "    \n",
    "    def run_burn2(self):\n",
    "        \"\"\"\n",
    "        burn to the point that all chromosomes are descended from a common ancestor\n",
    "        This sets the generation number at the base of the tree\n",
    "        total time will then be self.args.burn2_generation_time + self.args.treeDepth\n",
    "        \"\"\"\n",
    "        gen = 0\n",
    "        for i,c in enumerate(self.pop0):\n",
    "            c.ancestornumber = i\n",
    "        while self.pop0.checkancestors()== False:\n",
    "            self.pop0.generation(gen)\n",
    "            gen += 1        \n",
    "        return gen\n",
    "\n",
    "    def run(self):\n",
    "        \"\"\"\n",
    "        calls run_burn1(), run_burn2() and then  runs for treedepth generations\n",
    "        \"\"\"\n",
    "        self.args.burn1_generation_time,self.args.burn1_mean_fitness = self.run_burn1()\n",
    "        self.args.burn2_generation_time = self.run_burn2()\n",
    "        self.args.mutationexpectation_adjusted_for_burn2 = (self.args.mutationexpectation * self.args.treeDepth)/(self.args.treeDepth + self.args.burn2_generation_time)\n",
    "        allgen = self.args.burn2_generation_time\n",
    "        gen = 0\n",
    "        countpopgens = 0\n",
    "        self.pops['p1'] = self.pop0\n",
    "        while gen < self.args.treeDepth:\n",
    "            \"\"\"\n",
    "            loop over generations,  adding populations as needed\n",
    "            \"\"\"\n",
    "            if gen in self.times:\n",
    "                splitPop = self.split_generations[gen]\n",
    "                ## split populations\n",
    "                self.pops[splitPop[0]] = population(splitPop[0], self.pops[splitPop[1]], self.args)\n",
    "\n",
    "            for key in self.pops.keys():\n",
    "                if np.random.random() < self.args.adaptchangerate:\n",
    "                    aapos,newaa = self.pops[key].changeancestor()\n",
    "                numdifferentfitnessvalues = self.pops[key].generation(allgen)\n",
    "                self.args.meannumfits += numdifferentfitnessvalues\n",
    "                countpopgens +=1\n",
    "            \n",
    "            gen += 1\n",
    "            allgen += 1\n",
    "            if self.args.debug and gen % self.args.popsize2 == 0:\n",
    "                meanfits,meanmeanfit = self.fitCheck()\n",
    "                print(\"generation {} ({:.1f}%)  # populations: {} mean fitnesses: {} overall mean: {:.4f}  sample mutation counts: {}\".format(gen,100*gen/self.args.treeDepth,len(self.pops.keys()),meanfits,meanmeanfit,self.pop0[0].mcounts))\n",
    "                \n",
    "        self.args.meannumfits /= countpopgens\n",
    "        sample = self.samplefrompops()\n",
    "        # This does the magic to output population-level sequences\n",
    "        self.sampleindsfromonepop(self.args.nsampledindvs, self.args.fixpoptosamplefrom)\n",
    "        return sample\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parseargs():\n",
    "    parser = argparse.ArgumentParser(\"python makeSlimScript_JH.py\",formatter_class=argparse.ArgumentDefaultsHelpFormatter)\n",
    "    parser.add_argument(\"-A\", help=\"path for folder containing bacterial alignment files\",\n",
    "                        dest=\"bacaligndir\",required=True,type = str)\n",
    "    parser.add_argument(\"-b\", help=\"base filename for output files (name only, no directories)\",\n",
    "                        dest=\"basename\",required=True,type = str)\n",
    "    parser.add_argument(\"-d\", help=\"Debug mode. Generates screen output during run, with mutation counts by position at the end\",\n",
    "                        dest=\"debug\", default=False,action=\"store_true\")\n",
    "    parser.add_argument(\"-e\", help=\"random number seed for simulation (and for picking alignment if needed)\",\n",
    "                        dest=\"ranseed\",required=True,type=int)\n",
    "    parser.add_argument(\"-F\", help=\"directory path for output fasta file (default is same as for results file)\",\n",
    "                        dest=\"fdir\",type = str)\n",
    "    parser.add_argument(\"-g\", help=\"bacterial gene name, optional - if not used a random gene is selected\",\n",
    "                        dest=\"genename\",type=str)\n",
    "    parser.add_argument(\"-k\", help=\"Number of species (1, 4, 5 or 11)\",\n",
    "                        dest=\"numSpecies\",required=True,type=int)\n",
    "    parser.add_argument(\"-L\", help=\"Length of sequence (# amino acids)\",\n",
    "                        dest=\"aalength\",required=True,type=int)\n",
    "    parser.add_argument(\"-m\", help=\"Model file path\",\n",
    "                        dest=\"mssmodelfilename\",required = True,type = str)\n",
    "    parser.add_argument(\"-N\", help=\"Population size (diploid)\",\n",
    "                        dest=\"popsize\",required=True,type=int)\n",
    "    parser.add_argument(\"-o\", help=\"turn off a type of mutation: 1 for nonsynonymous, 2 for synonymous (requires -d))\",\n",
    "                        dest=\"turnoffmut\",default = 0, type=int)\n",
    "    parser.add_argument(\"-R\", help=\"directory path for results file\",\n",
    "                        dest=\"rdir\",default = \".\",type = str)\n",
    "    parser.add_argument(\"-q\", help=\"compress/expand run time, useful when debugging to get a short run,requires -d\",\n",
    "                        dest=\"treerescaler\",default = 1.0, type=float)\n",
    "    parser.add_argument(\"-s\", help=\"Synonymous population selection coefficient, 2Ns (Slim uses 1-(2Ns/2N))\",\n",
    "                        dest=\"SynSel_s\",default=2,type=float)\n",
    "    parser.add_argument(\"-y\", help=\"Non-synonymous population selection coefficient, 2Ns (Slim uses 1-(2Ns/2N))\",\n",
    "                        dest=\"NonSyn_s\",default=10,type=float)\n",
    "    parser.add_argument(\"-u\", help=\"expected number of neutral mutations per site, from base of tree\",\n",
    "                        dest=\"mutationexpectation\",default=0.5,type=float)\n",
    "    parser.add_argument(\"-w\", help=\"rate per generation of adaptive amino acid change\",\n",
    "                        dest=\"adaptchangerate\",default=0.0,type=float)\n",
    "    parser.add_argument(\"-x\", help=\"save detailed substitution info (requires -d, slows things down a good bit)\",\n",
    "                        dest=\"savesubtimeinfo\",default = False, action=\"store_true\")\n",
    "    parser.add_argument(\"-n\", help=\"number of sampled individuals to save to population-level alignement file\",\n",
    "                        dest=\"nsampledindvs\",default = 0, type=int)\n",
    "    parser.add_argument(\"-p\", help=\"the population to sample individuals from (sp <= k)\",\n",
    "                        dest=\"fixpoptosamplefrom\",default = 0, type=int)\n",
    "    return parser"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Code below is to run iteratively (to learn what each command does.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#argv = \"-A ../work/bacalignments/ -b xx -e 2 -F ../work/runs/debug -k 5 -L 300 -m MSS_MULTIPLE_NEUTRAL_SETS.tsv -N 100 -R ../work/runs/debug -s 0.5 -y 3 -u 10 -w 0 -q 0.1 -d\"\n",
    "# argv = \"-A ../work/bacalignments/ -b xx -e 2 -F ../work/runs/debug -k 5 -L 300 -m MSS_MULTIPLE_NEUTRAL_SETS.tsv -N 100 -R ../work/runs/debug -s 0.5 -y 3 -u 10 -w 0 -q 0.1 -d\"\n",
    "\n",
    "# argv = argv.split()\n",
    "# print(argv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# global args\n",
    "\n",
    "# starttime = time.time()\n",
    "# parser = parseargs()\n",
    "# if argv[-1] =='':\n",
    "#     argv = argv[0:-1]\n",
    "# args = parser.parse_args(argv)\n",
    "# if args.ranseed != None:\n",
    "#     np.random.seed(args.ranseed)\n",
    "# if args.numSpecies not in [1, 4,5,11]:\n",
    "#     print (\"error: -p (# of species) must be 1, 4,5 or 11\")\n",
    "#     exit()\n",
    "# args.commandstring = \" \".join(argv)\n",
    "# if args.debug:\n",
    "#     args.debug = debuginfo(args.turnoffmut != 0,args.savesubtimeinfo == True,args.treerescaler != 1.0)\n",
    "# else: # args.debug==False\n",
    "#     if args.turnoffmut != 0:\n",
    "#         args.commandstring += \"\\n\\t -o ignored \"\n",
    "#         print(\"-o ignored\")\n",
    "#         args.turnoffmut = 0\n",
    "#     if args.treerescaler != 1.0:\n",
    "#         print(\"-q ignored\")\n",
    "#         args.commandstring += \"\\n\\t -q ignored \"\n",
    "#         args.treerescaler = 1.0\n",
    "#     if args.savesubtimeinfo:\n",
    "#         print(\"-x ignored\")\n",
    "#         args.commandstring += \"\\n\\t -x ignored \"\n",
    "#         args.savesubtimeinfo = False \n",
    "\n",
    "# args.meannumfits = 0\n",
    "# args.popsize2 = args.popsize*2\n",
    "# args.defaulttreeDepth = 100000 # fixed at a specific value # previously scaled by population size  args.treeDepth * args.popsize\n",
    "# args.mutrate = args.mutationexpectation/args.defaulttreeDepth  # got rid of using theta 4Nu,  as not really relevant here \n",
    "# args.treeDepth = round(args.defaulttreeDepth * args.treerescaler)\n",
    "# if args.debug and args.debug.userescaler:\n",
    "#     args.treeDepth = round(args.defaulttreeDepth * args.treerescaler)\n",
    "# else:\n",
    "#     args.treeDepth = args.defaulttreeDepth\n",
    "# #rescale the selection coefficients from 2Ns values to Slim values\n",
    "# args.SynSelDel_s_rescaled = max(0.0,1.0 - (args.SynSel_s/(args.popsize2)))\n",
    "\n",
    "# args.SynSelFav_s_rescaled = 1.0/args.SynSelDel_s_rescaled  # makes more sense to use the reciprocal for a favored change\n",
    "# # args.SynSelFav_s_rescaled = 1.0 + (1.0 - args.SynSelDel_s_rescaled)\n",
    "\n",
    "# args.NonSyn_s_rescaled = max(0.0,1.0 - (args.NonSyn_s/(args.popsize2)))\n",
    "# if args.NonSyn_s_rescaled <= 0.0:\n",
    "#     print(\"fitness error\")\n",
    "#     exit()\n",
    "# curdir = os.getcwd()\n",
    "# #if path is a string that can be separated into folders,  and one or more of them do not exist,  this will create them\n",
    "# try:# create folder(s) if needed. When running lots of jobs, a dir may not exist at one moment, but then does exist in the next because of another job running, so use try/except\n",
    "#     curdir = os.getcwd()\n",
    "#     normalized_path = op.normpath(args.rdir)\n",
    "#     dirs = normalized_path.split(os.sep)\n",
    "#     for d in dirs:\n",
    "#         if op.exists(d) == False:\n",
    "#             os.mkdir(d)\n",
    "#         os.chdir(d)\n",
    "#     os.chdir(curdir)\n",
    "# except:\n",
    "#     pass\n",
    "# if args.fdir== None:\n",
    "#     args.fdir=args.rdir\n",
    "# else:\n",
    "#     #if path is a string that can be separated into folders,  and one or more of them do not exist,  this will create them\n",
    "#     try:# create folder(s) if needed. When running lots of jobs, a dir may not exist at one moment, but then does exist in the next because of another job running, so use try/except\n",
    "#         curdir = os.getcwd()\n",
    "#         normalized_path = op.normpath(args.fdir)\n",
    "#         dirs = normalized_path.split(os.sep)\n",
    "#         for d in dirs:\n",
    "#             if op.exists(d) == False:\n",
    "#                 os.mkdir(d)\n",
    "#             os.chdir(d)\n",
    "#         os.chdir(curdir)        \n",
    "#     except:\n",
    "#         pass  \n",
    "    \n",
    "# #create global vars\n",
    "# global mainmutationcounter #when debug update this when mutating\n",
    "# global nummutationtypes\n",
    "# global  NonSynDelX,  NonSynFavX, SynSelX,  SynNeuX, STOPX # all refer to positions in the mutation counter arrays,  all end in 'X' \n",
    "\n",
    "# nummutationtypes = 5\n",
    "# #positions in mutation counter lists\n",
    "# NonSynDelX = 0 # nonsynonymous deleterious\n",
    "# NonSynFavX = 1 # nonsynonymous favored (0's unless adaptive changes can occur via -w  )\n",
    "# SynSelX = 2  # synonymous selected \n",
    "# SynNeuX = 3  # synonymous neutral\n",
    "# STOPX = 4  # nonsense \n",
    "# global codondic, aalist, codonlist, revcodondic,stopCodons\n",
    "# codondic, aalist, codonlist, revcodondic,stopCodons = codonInfo()\n",
    "# mainmutationcounter = [0 for i in range(nummutationtypes)]  #positions 0,1,2 or 3 for nonsynonymous,  synonymous-selected, synonymous-neutral, and STOP \n",
    "# global mutationlocations # used when debugging for checking distribution of mutation locations\n",
    "# mutationlocations = [0 for i in range(3*args.aalength)]\n",
    "\n",
    "# # get ancestral sequence\n",
    "# args.fitnessstructure, args.mutstructure,args.neutralsets,args.codon_substitution_time_info_dict = createSelectedDictionary(args)\n",
    "# args.ancestor,args.ancestorfitness,genefilename = makeAncestor(args)\n",
    "# args.genename = genefilename[:genefilename.find('_')]\n",
    "\n",
    "# args.resultsfilename = op.join(args.rdir,args.basename +  \"_\" + args.genename + '_results.txt')\n",
    "# if os.path.exists(args.resultsfilename):\n",
    "#     args.resultsfilename = '{}(1)_results.txt'.format(args.resultsfilename[:-12])\n",
    "# args.fastafilename = op.join(args.fdir, args.basename +  \"_\" + args.genename +\".fa\")\n",
    "# if os.path.exists(args.fastafilename):\n",
    "#     args.fastafilename = '{}(1).fa'.format(args.fastafilename[:-3])\n",
    "\n",
    "# # set tree shape\n",
    "# args.tree, args.split_generations,args.mean_branches_root_to_tip = maketreeshape(args.numSpecies)\n",
    "\n",
    "# # run the simulation\n",
    "# sim = tree(args)\n",
    "# sampledsequences = sim.run()\n",
    "# sim.summarize_results(starttime,sampledsequences)\n",
    "# if args.debug:\n",
    "#     print(\"mutation counts by base position\\n\",mutationlocations)\n",
    "# makefastafile(sampledsequences, args.fastafilename)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Code below is to run as the simulator was designed to run, by parsing command line arguments (see next cells)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [],
   "source": [
    "def main(argv):\n",
    "    global args\n",
    "\n",
    "    starttime = time.time()\n",
    "    parser = parseargs()\n",
    "    if argv[-1] =='':\n",
    "        argv = argv[0:-1]\n",
    "    args = parser.parse_args(argv)\n",
    "    if args.ranseed != None:\n",
    "        np.random.seed(args.ranseed)\n",
    "    if args.numSpecies not in [1, 4,5,11]:\n",
    "        print (\"error: -p (# of species) must be 1, 4,5 or 11\")\n",
    "        exit()\n",
    "    args.commandstring = \" \".join(argv)\n",
    "    if args.debug:\n",
    "        args.debug = debuginfo(args.turnoffmut != 0,args.savesubtimeinfo == True,args.treerescaler != 1.0)\n",
    "    else: # args.debug==False\n",
    "        if args.turnoffmut != 0:\n",
    "            args.commandstring += \"\\n\\t -o ignored \"\n",
    "            print(\"-o ignored\")\n",
    "            args.turnoffmut = 0\n",
    "        if args.treerescaler != 1.0:\n",
    "            print(\"-q ignored\")\n",
    "            args.commandstring += \"\\n\\t -q ignored \"\n",
    "            args.treerescaler = 1.0\n",
    "        if args.savesubtimeinfo:\n",
    "            print(\"-x ignored\")\n",
    "            args.commandstring += \"\\n\\t -x ignored \"\n",
    "            args.savesubtimeinfo = False \n",
    "    \n",
    "    args.meannumfits = 0\n",
    "    args.popsize2 = args.popsize*2\n",
    "    args.defaulttreeDepth = 100000 # fixed at a specific value # previously scaled by population size  args.treeDepth * args.popsize\n",
    "    args.mutrate = args.mutationexpectation/args.defaulttreeDepth  # got rid of using theta 4Nu,  as not really relevant here \n",
    "    args.treeDepth = round(args.defaulttreeDepth * args.treerescaler)\n",
    "    if args.debug and args.debug.userescaler:\n",
    "        args.treeDepth = round(args.defaulttreeDepth * args.treerescaler)\n",
    "    else:\n",
    "        args.treeDepth = args.defaulttreeDepth\n",
    "    #rescale the selection coefficients from 2Ns values to Slim values\n",
    "    args.SynSelDel_s_rescaled = max(0.0,1.0 - (args.SynSel_s/(args.popsize2)))\n",
    "    \n",
    "    args.SynSelFav_s_rescaled = 1.0/args.SynSelDel_s_rescaled  # makes more sense to use the reciprocal for a favored change\n",
    "    # args.SynSelFav_s_rescaled = 1.0 + (1.0 - args.SynSelDel_s_rescaled)\n",
    "    \n",
    "    args.NonSyn_s_rescaled = max(0.0,1.0 - (args.NonSyn_s/(args.popsize2)))\n",
    "    if args.NonSyn_s_rescaled <= 0.0:\n",
    "        print(\"fitness error\")\n",
    "        exit()\n",
    "    curdir = os.getcwd()\n",
    "    #if path is a string that can be separated into folders,  and one or more of them do not exist,  this will create them\n",
    "    try:# create folder(s) if needed. When running lots of jobs, a dir may not exist at one moment, but then does exist in the next because of another job running, so use try/except\n",
    "        curdir = os.getcwd()\n",
    "        normalized_path = op.normpath(args.rdir)\n",
    "        dirs = normalized_path.split(os.sep)\n",
    "        for d in dirs:\n",
    "            if op.exists(d) == False:\n",
    "                os.mkdir(d)\n",
    "            os.chdir(d)\n",
    "        os.chdir(curdir)\n",
    "    except:\n",
    "        pass\n",
    "    if args.fdir== None:\n",
    "        args.fdir=args.rdir\n",
    "    else:\n",
    "        #if path is a string that can be separated into folders,  and one or more of them do not exist,  this will create them\n",
    "        try:# create folder(s) if needed. When running lots of jobs, a dir may not exist at one moment, but then does exist in the next because of another job running, so use try/except\n",
    "            curdir = os.getcwd()\n",
    "            normalized_path = op.normpath(args.fdir)\n",
    "            dirs = normalized_path.split(os.sep)\n",
    "            for d in dirs:\n",
    "                if op.exists(d) == False:\n",
    "                    os.mkdir(d)\n",
    "                os.chdir(d)\n",
    "            os.chdir(curdir)        \n",
    "        except:\n",
    "            pass  \n",
    "        \n",
    "    #create global vars\n",
    "    global mainmutationcounter #when debug update this when mutating\n",
    "    global nummutationtypes\n",
    "    global  NonSynDelX,  NonSynFavX, SynSelX,  SynNeuX, STOPX # all refer to positions in the mutation counter arrays,  all end in 'X' \n",
    "    \n",
    "    nummutationtypes = 5\n",
    "    #positions in mutation counter lists\n",
    "    NonSynDelX = 0 # nonsynonymous deleterious\n",
    "    NonSynFavX = 1 # nonsynonymous favored (0's unless adaptive changes can occur via -w  )\n",
    "    SynSelX = 2  # synonymous selected \n",
    "    SynNeuX = 3  # synonymous neutral\n",
    "    STOPX = 4  # nonsense \n",
    "    global codondic, aalist, codonlist, revcodondic,stopCodons\n",
    "    codondic, aalist, codonlist, revcodondic,stopCodons = codonInfo()\n",
    "    mainmutationcounter = [0 for i in range(nummutationtypes)]  #positions 0,1,2 or 3 for nonsynonymous,  synonymous-selected, synonymous-neutral, and STOP \n",
    "    global mutationlocations # used when debugging for checking distribution of mutation locations\n",
    "    mutationlocations = [0 for i in range(3*args.aalength)]\n",
    "\n",
    "    # get ancestral sequence\n",
    "    \n",
    "    args.fitnessstructure, args.mutstructure,args.neutralsets,args.codon_substitution_time_info_dict = createSelectedDictionary(args)\n",
    "    args.ancestor,args.ancestorfitness,genefilename = makeAncestor(args)\n",
    "    args.genename = genefilename[:genefilename.find('_')]\n",
    "\n",
    "    args.resultsfilename = op.join(args.rdir,args.basename +  \"_\" + args.genename + '_results.txt')\n",
    "    if os.path.exists(args.resultsfilename):\n",
    "        args.resultsfilename = '{}(1)_results.txt'.format(args.resultsfilename[:-12])\n",
    "    args.fastafilename = op.join(args.fdir, args.basename +  \"_\" + args.genename +\".fa\")\n",
    "    if os.path.exists(args.fastafilename):\n",
    "        args.fastafilename = '{}(1).fa'.format(args.fastafilename[:-3])\n",
    "\n",
    "    # set tree shape\n",
    "    args.tree, args.split_generations,args.mean_branches_root_to_tip = maketreeshape(args.numSpecies)\n",
    "\n",
    "    # run the simulation\n",
    "    sim = tree(args)\n",
    "    sampledsequences = sim.run()\n",
    "    sim.summarize_results(starttime,sampledsequences)\n",
    "    if args.debug:\n",
    "        print(\"mutation counts by base position\\n\",mutationlocations)\n",
    "    makefastafile(sampledsequences, args.fastafilename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "# argv = \"-A ../work/bacalignments/ -b xx -e 2 -F ../work/runs/debug -k 4 -L 300 -m MSS_MULTIPLE_NEUTRAL_SETS.tsv -N 100 -R ../work/runs/debug -s 0.5 -y 3 -u 10 -w 0 -q 0.1 -d\"\n",
    "# sys.argv = argv.split()\n",
    "# print(sys.argv)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "if __name__ == \"__main__\":\n",
    "\n",
    "    if len(sys.argv) < 2:\n",
    "        main(['-h'])\n",
    "    else:\n",
    "        #main(sys.argv[1:])\n",
    "        main(sys.argv)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mssFsim",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
